src/v/coproc/partition.cc

```diff
@@ -0,0 +1,100 @@
+/*
+ * Copyright 2021 Vectorized, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#include "coproc/partition.h"
+
+#include "storage/log.h"
+
+namespace {
+
+/// Ensures that operations on a model::record_batch_reader occur within the
+/// context of a gate for safe shutdown. If operations are observed while the
+/// gate is closed the caller will observe the ss::gate_closed_exception
+class gated_shutdown_reader final : public model::record_batch_reader::impl {
+public:
+    gated_shutdown_reader(
+      ss::gate& g,
+      std::unique_ptr<model::record_batch_reader::impl> impl) noexcept
+      : _gate(g)
+      , _impl(std::move(impl)) {}
+
+    bool is_end_of_stream() const final {
+        return _gate.is_closed() || _impl->is_end_of_stream();
+    }
+
+    ss::future<model::record_batch_reader::storage_t>
+    do_load_slice(model::timeout_clock::time_point t) final {
+        return ss::with_gate(
+          _gate, [this, t] { return _impl->do_load_slice(t); });
+    }
+
+    void print(std::ostream& os) final { os << "gated_shutdown_reader"; }
+
+private:
+    ss::gate& _gate;
+    std::unique_ptr<model::record_batch_reader::impl> _impl;
+};
+
+} // namespace
+
+namespace coproc {
+/// Not neccessary to hold gate open until all processing completes, as that can
+/// be done by invoking the write within the context of the \ref
+/// non_replicable_partition gate. This class just makes the operation
+/// cancellable
+safe_shutdown_appender::safe_shutdown_appender(
```

Comment:
if it will be possible to make sure that `log` lives longer than `appender` this will not be needed..

Reply:
Yes this is true, however to do this in coproc wouldn't be immediately trivial w.r.t. materialized topic deletion and in-progress writes from fibers. An idea i have is having a temporary list of `to_be_deleted` topics, and scripts can refer to that and not write to that list while there is an in-progress delete. There will need to be some code to synchronize that the `reconciliation_backend` doesn't progress until all coproc fibers are respecting the update to this list.

I am weighing a change like that with the alternative of something like this, which will keep the gate open until there are no more active writers. Thoughts?

Reply:
I don't understand why we can call gate.close on an appender object to make shutdown safe, but we can't keep the log alive long enough to accomplish the same safety.

Reply:
gate.close() is never called on the appender. The appender holds a reference to the gate owned by the `coproc::partition`. The idea is the partition is kept alive long enough that we achieve 'safety'

Reply:
Do I understand correctly that you are using your safe_appender as a vector to inject a stop-signal into the fiber that is appending?

Reply:
Exactly. If a fiber attempts to write after the stop signal, they will receive `gate_closed_exception`, and this is handled in the last commit in this PR

Reply:
so since you control everything, why don't you inject the stop signal into the fiber?

Reply:
Well this is kind of doing that, since I own `coproc::partition`. Unless i'm not thinking of some other way this could be done?

Reply:
safe_appender is an indirect signal, but a gate or an abort source can be added to the fiber as a means of directly representing the signal.

there is nothing inherently wrong with injecting the signal through the appender, but as we have seen, it's not an expected design pattern. at the very least I think you need to document in this in a way that a new developers or some debugging an issue would be able to avoid this discussion we've had to establish what is happening.

Reply:
Ok i understand your point now. So I have another idea, where I pause and resume fibers using the direct signal mechanisms your mentioning above, I would just need to add this mechanism into the fibers themselves, was weighing this additional complexity against the `safe_appender`

Reply:
it's really up to you. 90% of this was understanding what was happening. the remaining 10% should be to document this for everyone else, and decide if it's a design pattern you want to maintain.